### PR TITLE
Reworked to mine Quai

### PR DESCRIPTION
**Pull request by Djadih**
Started off Alephium's gpu-miner and reworked the codebase to be compatible with our miner. This includes:
- improved network performance: mining with a sealhash prepended with nonce, rather than the miner sending the whole raw header itself
- modularity: changes to the header no longer impact the miner implementation
- 2x hash rate: performing a single hash instead of double hash like alephium

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -41,6 +41,7 @@ conan_cmake_run(REQUIRES libuv/1.42.0
 # Target definitions
 add_executable(gpu-miner src/main.cu)
 set_property(TARGET gpu-miner PROPERTY CXX_STANDARD 11)
+# set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} --ptxas-options=-v --debug -g -G -lineinfo")        # For debugging
```
Comment (gameofpointers) [inline CMakeLists.txt:44]: why is this commented? isn't there a nicer way to do this!!

Comment (wizeguyy) [reply CMakeLists.txt:44]: +1 if we don't need it, remove it. If we conditionally need it, make it switch on a flag.

File: src/main.cu [hunk 2]
```diff
@@ -35,7 +35,7 @@ std::atomic<uint64_t> total_mining_count;
 std::atomic<uint64_t> device_mining_count[max_gpu_num];
 bool use_device[max_gpu_num];
 
-int port = 10973;
+int port = 8008;
```
Comment (wizeguyy) [inline src/main.cu:38]: Please use a common stratum port number

File: src/messages.h [hunk 3]
```diff
@@ -73,34 +79,24 @@ void hex_to_bytes(const char *hex_data, blob_t *buf)
...
+    char* header_msg;
+} header_msg_t;
+
+// job struct
 typedef struct job_t {
-    int from_group;
-    int to_group;
     blob_t header_blob;
-    blob_t txs_blob;
+    // blob_t txs_blob;
```
Comment (wizeguyy) [inline src/messages.h]: Please remove dead code from this commit

File: src/main.cu [hunk 4]
```diff
@@ -115,19 +115,16 @@ void mine(mining_worker_t *worker)
...
     } else {
-        mining_counts[to_mine_index].fetch_add(mining_steps);
-        setup_template(worker, load_template(to_mine_index));
-
+        mining_count.fetch_add(mining_steps);
+        setup_template(worker, load_template());
         start_worker_mining(worker);
 
-        duration_t elapsed = Time::now() - start;
+        // duration_t elapsed = Time::now() - start;
```
Comment (wizeguyy) [inline src/main.cu]: More dead code to remove

**Review by wizeguyy — APPROVED** [review]
A few minor comments, but over all looks good!

File: CMakeLists.txt (hunk 1 above)
Comment (Djadih) [reply CMakeLists.txt:44]: I'll move it to the end of the line, just so others can find it if they want to debug. I personally didn't know these flags existed and was missing some which made debugging difficult.

File: src/main.cu (hunk 2 above)
Comment (Djadih) [reply src/main.cu:38]: Good idea, 3333 appears to be the standard
https://help.braiins.com/en/support/solutions/articles/77000423438-bitcoin-mining-urls-stratum-v1-

**Comment by Djadih** [comment]
Results of running 4 day test. All miners still running with 4 threads each.
image

**Review by wizeguyy — APPROVED** [review]

**Pull request merged**